test/test_testing.py

```diff
@@ -744,413 +745,396 @@ def test_trivial_passing_test(self, device):
                 fn()
 
-    @onlyCPU
-    def test_unknown_type(self, device):
+    def test_unknown_type(self):
         actual = "0"
         expected = "0"
 
-        for fn in self.assert_fns_with_inputs(actual, expected):
+        for fn in assert_fns_with_inputs(actual, expected):
             with self.assertRaisesRegex(UsageError, str(type(actual))):
                 fn()
 
-    @onlyCPU
-    def test_numpy(self, device):
-        tensor = torch.rand(2, 2, dtype=torch.float32, device=device)
+    def test_numpy(self):
+        tensor = torch.rand(2, 2, dtype=torch.float32)
         actual = tensor.numpy()
         expected = actual.copy()
 
-        for fn in self.assert_fns_with_inputs(actual, expected):
+        for fn in assert_fns_with_inputs(actual, expected):
             fn()
 
-    @onlyCPU
-    def test_scalar(self, device):
-        number = torch.randint(10, size=(), device=device).item()
+    def test_scalar(self):
+        number = torch.randint(10, size=()).item()
         for actual, expected in itertools.product((int(number), float(number), complex(number)), repeat=2):
             check_dtype = type(actual) is type(expected)
 
-            for fn in self.assert_fns_with_inputs(actual, expected):
+            for fn in assert_fns_with_inputs(actual, expected):
                 fn(check_dtype=check_dtype)
 
-    @onlyCPU
-    def test_msg_str(self, device):
+    def test_msg_str(self):
         msg = "Custom error message!"
 
-        actual = torch.tensor(1, device=device)
-        expected = torch.tensor(2, device=device)
+        actual = torch.tensor(1)
+        expected = torch.tensor(2)
 
-        for fn in self.assert_fns_with_inputs(actual, expected):
+        for fn in assert_fns_with_inputs(actual, expected):
             with self.assertRaisesRegex(AssertionError, msg):
                 fn(msg=msg)
 
-    @onlyCPU
-    def test_msg_callable(self, device):
+    def test_msg_callable(self):
         msg = "Custom error message!"
 
         def make_msg(actual, expected, trace):
             return msg
 
-        actual = torch.tensor(1, device=device)
-        expected = torch.tensor(2, device=device)
+        actual = torch.tensor(1)
+        expected = torch.tensor(2)
 
-        for fn in self.assert_fns_with_inputs(actual, expected):
+        for fn in assert_fns_with_inputs(actual, expected):
             with self.assertRaisesRegex(AssertionError, msg):
                 fn(msg=make_msg)
 
 
-instantiate_device_type_tests(TestAsserts, globals())
+def device_permutations(cuda_devices: List[str]) -> Iterator[Tuple[str, str]]:
+    """Yields all possible permutations of the CPU and the passed CUDA devices.
+
+    Yields:
+        Tuple[str, str]: Device combinations
+    """
+    if not cuda_devices:
+        return
+
+    yield from itertools.permutations(("cpu", *cuda_devices), 2)
+
+
+class TestAssertsMultiDevice(TestCase):
+    @deviceCountAtLeast(1)
```

Comment:
@mruberry I dislike this for two reasons:

1. We now need to slap the same decorator on every test in this test case.
2. Without knowing that `@deviceCountAtLeast(1)` returns a list of available devices rather than a single device this can be confusing: 
    > Why do I need this decorator if the test case is instatiated with `only_for="cuda"`? Doesn't that already check that CUDA is available and thus I have at least one device?"

Reply:
It lets you get rid of the `device_permutations` helper because you can inline it as just

```
itertools.permutations(("cpu", *cuda_devices), 2)
```

Reply:
Looking forward, too, it's important today that we understand if a test would use a hardware accelerator or not, tomorrow it might be important we understand if it could use multiple hardware accelerators, and this is decorator is the best way we have of understanding that programmatically